map.go

```diff
@@ -1622,12 +1649,56 @@ func newMapIterator(target *Map) *MapIterator {
 	}
 }
 
+// cursorToKeyOut copies the current value held in the cursor to the
+// provided argument. In case of errors, returns false and sets a
+// non-nil error in the MapIterator.
+func (mi *MapIterator) cursorToKeyOut(keyOut interface{}) bool {
+	buf := mi.cursor.([]byte)
+	if ptr, ok := keyOut.(unsafe.Pointer); ok {
+		copy(unsafe.Slice((*byte)(ptr), len(buf)), buf)
+	} else {
+		mi.err = sysenc.Unmarshal(keyOut, buf)
+	}
+	return mi.err == nil
+}
+
+// fetchNextKey loads into the cursor the key following the provided one.
+func (mi *MapIterator) fetchNextKey(key interface{}) bool {
+	mi.err = mi.target.NextKey(key, mi.cursor)
+	if mi.err == nil {
+		return true
+	}
+
+	if errors.Is(mi.err, ErrKeyNotExist) {
+		mi.done = true
+		mi.err = nil
+	} else {
+		mi.err = fmt.Errorf("get next key: %w", mi.err)
+	}
+
+	return false
+}
+
+// drainMapEntry removes and returns the key held in the cursor
+// from the underlying map.
+func (mi *MapIterator) drainMapEntry(valueOut interface{}) bool {
+	mi.err = mi.target.LookupAndDelete(mi.cursor, valueOut)
+	if mi.err == nil {
+		mi.count++
+		return true
+	}
+
+	if errors.Is(mi.err, ErrKeyNotExist) {
+		mi.err = nil
+	} else {
+		mi.err = fmt.Errorf("lookup_and_delete key: %w", mi.err)
```

Comment:
Prefer natural language in error strings, these will be seen by the user:
```suggestion
		mi.err = fmt.Errorf("lookup and delete next key: %w", mi.err)
```